Introducing an implementation for the UTT signature collector

yontyon opened the pull request:
* **Problem Overview**  
  In this PR we introduce the basic implementation of the UTT signature processor component.
  The signature processor is responsible for computing the partial signature, collecting others' partial signatures, assembling the full signature, and publishing it back through a consensus to the system.
  
  In this implementation, we introduce a leader-based approach that relies on the bft consensus for correctness and safety.
  Also, the signature processor is thread-safe such that the entry point method can be called by any thread.

When handling failures (i.e. the full valid signature was not received by consensus in time) we go over the jobs and retransmit every overdue job to all. However, if there are many jobs this may take a while. For that, we also save the partial signatures in an ordered map (ordered by the overdue date of the job) which allows us to go only over the jobs which exceeded their timeout.
  
* **Testing Done**  
    We also added some simple unittests that demonstrate how to use the signature processor

teoparvanov commented on utt-replica/signature-processor/tests/sigProcessorTests.cpp:
```diff
@@ -0,0 +1,519 @@
+uint16_t n = 4;
+uint16_t t = 3;
+
+int main(int argc, char* argv[]) {
+  // initiate A UTT instance utt1
```
Ideally this test should be implemented as a gtest suite? For now we can track it as a tech debt item for faster progress.

yontyon commented on utt-replica/signature-processor/include/sigData.hpp:
```diff
@@ -0,0 +1,87 @@
+class SigJobEntriesMap {
+ public:
+  const SigJobEntry& get(uint64_t id) const {
+    std::unique_lock lck(lock_);
```
here we also return const

evdzhurov commented on utt-replica/signature-processor/src/sigProcessor.cpp:
```diff
@@ -0,0 +1,157 @@
+                                    const std::vector<uint8_t>& sig,
+                                    const validationCb& vcb,
+                                    uint64_t job_timeout_ms) {
+  jobs_.registerEntry(sig_id, vcb, job_timeout_ms);
+  const auto& jobEntry = jobs_.get(sig_id);
+  auto& validator = jobs_.getJobValidator(sig_id);
+  std::vector<uint16_t> invalids;
+  for (auto& [id, psig] : jobEntry.partial_sigs) {
+    if (!validator(id, psig)) invalids.push_back((uint16_t)id);
+  }
+  for (auto id : invalids) jobs_.removePartialSig(sig_id, (uint32_t)id);
+
+  jobs_.addPartialSig(sig_id, (uint32_t)repId_, sig);
+  auto job_coordinator = get_job_coordinator_cb(sig_id);
+  if (job_coordinator == repId_) return;
+  messages::PartialSigMsg msg(repId_, sig, sig_id);
+  msgs_communicator_->sendAsyncMessage((uint64_t)job_coordinator, msg.body(), msg.size());
```
This is a bit of mess. You get a jobEntry const ref that you read from but at the same time you modify it multiple times through other SigJobEntriesMap methods. Maybe if you try to get rid of SigJobEntriesMap and do the locking in sig processor you will end up with much simpler and straightforward code.

yontyon replied:
The main reason for introducing the SigJobEntriesMap is to have the sigProcessor code focused only on the sigProcessor logic (while introducing a fine granularity locking as possible).
As for the specific comment, we don't change the entry while reading it, we only get the const reference and validate the signatures outside of the lock (while letting other threads add/remove data to the complete job maps).  

of course, all of this can be done inside the code of the sigPorcessor, but as I said, I personally find it easier to focus on the signature collecting process only here and leave the storage/locking and all the dirty things to some another internal component

yontyon replied:
Basically, the challenge here is not to block the other threads while validating the partial signatures.
Now, recall that SigJobEntriesMap is not supposed to be a complete thread-safe data structure, but more of a specific implementation of thread-safe container for the signature collector.
For example, we do know that an honest replica would never change the entry after it was registered.
We also know who are the callers of the methods (the signer threads and the message processing thread).
I'm not sure it worth investing in building a complete thread-safe data structure, and locking in the sigProcessor level, will probably introduce similar issues if we want to keep the fine granularity of the locking... 
I will try to design maybe a more comprehensible data container layer.
Anyway, regarding the comment itself, we don't change the entry while reading it, we first read it (and let other threads to grab the lock), then we accumulate the invalid signatures, then we grab the lock again and remove the invalid signatures.
Another maybe straightforward solution is to add a mutex per entry, but again, I think that considering the involved threads and the data flow, it might quite overkill.
WDYT?

yontyon commented on bftengine/CMakeLists.txt:
```diff
@@ -89,6 +89,12 @@ add_library(bftclient STATIC src/bftengine/SimpleClientImp)
+    target_compile_options(bftheaders INTERFACE -Wno-sign-conversion)
+    target_compile_options(corebft PRIVATE -Wno-sign-conversion)
+endif (HAVE_SIGN_CONVERSION)
+
```
fixed

yontyon commented on utt-replica/signature-processor/CMakeLists.txt:
```diff
@@ -0,0 +1,43 @@
+
+set(sig_processor_src src/sigProcessor.cpp)
+add_library(sig_processor STATIC ${sig_processor_src})
+target_link_libraries(sig_processor PUBLIC utt_api corebft util)
```
I'm not sure about this for two reasons:
1. The only reason we need libutt is for aggregating signatures, and aggregating signatures is not necessarily part of utt-server-api
2.  The usage of both, sigProcessor and libutt inside it is completely internal, so why not use libutt? it has a public interface, why add another layer in between?

yontyon commented on utt-replica/signature-processor/include/sigData.hpp:
```diff
@@ -0,0 +1,87 @@
+
+  const std::unordered_map<uint64_t, SigJobEntry>& getEntries() {
+    std::unique_lock lck(lock_);
+    return jobs_;
```
fixed by making the method const

yontyon commented:
@evdzhurov , I've fixed the thread-safe issues, please review again and let me know if I missed anything.
@cloudnoize , I've addressed your concern about blocking the message processor thread on timeout. By maintaining an ordered map by the tasks' due-date we simply can go only over the overdue tasks and retransmit them, please review again.

Thanks

evdzhurov reviewed (approved):
LGTM

teoparvanov reviewed (approved):
Thanks for addressing the reviewers' feedback!
